Improving the default PDB implementation.

naemono opened the pull request:
Resolves #2936 

## What is this change?

This adds an Enterprise feature which allows the ECK operator to create multiple PDBs which map to grouped roles within nodesets (ideally a PDB per nodeset of role 'X'). For non-enterprise customers, the behavior remains the same, which creates a single default pdb for the whole cluster allowing 1 disruption when the cluster is green.

### Examples

Simple example (pdb per nodeset with single roles)
```
# Manifest
kind: Elasticsearch
spec:
  version: 8.12.1
  nodeSets:
    - name: masters
      count: 3
      config:
        node.roles: ["master"]
    - name: data
      count: 3
      config:
        node.roles: ["data", "ingest"]

# PDBs created
apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  labels:
    common.k8s.elastic.co/type: elasticsearch
    elasticsearch.k8s.elastic.co/cluster-name: cluster-name
  name: cluster-name-es-default-master
spec:
  maxUnavailable: 1
  selector:
    matchExpressions:
    -  key: elasticsearch.k8s.elastic.co/cluster-name
        operator: "In"
        values: ["cluster-name"]
    -  key: elasticsearch.k8s.elastic.co/statefulset-name
        operator: "In"
        values: ["cluster-name-es-masters"]

apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  labels:
    common.k8s.elastic.co/type: elasticsearch
    elasticsearch.k8s.elastic.co/cluster-name: cluster-name
  name: cluster-name-es-default-data
spec:
  maxUnavailable: 1
  selector:
    matchExpressions:
    -  key: elasticsearch.k8s.elastic.co/cluster-name
        operator: "In"
        values: ["cluster-name"]
    -  key: elasticsearch.k8s.elastic.co/statefulset-name
        operator: "In"
        values: ["cluster-name-es-data"]
```

Complex example (mixed roles across nodeSets)
```
# Manifest
kind: Elasticsearch
spec:
  version: 8.12.1
  nodeSets:
    - name: masters
      count: 3
      config:
        node.roles: ["master", "data", "ingest"]
    - name: data-warm
      count: 3
      config:
        node.roles: ["data_warm"]
    - name: data-cold
      count: 3
      config:
        node.roles: ["data_cold"]
    - name: data-frozen
      count: 3
      config:
        node.roles: ["data_frozen"]
    - name: coordinating
      count: 1
      config:
        node.roles: []

# PDBs created
apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  labels:
    common.k8s.elastic.co/type: elasticsearch
    elasticsearch.k8s.elastic.co/cluster-name: cluster-name
  name: cluster-name-es-default-data
spec:
  maxUnavailable: 1
  selector:
    matchExpressions:
    -  key: elasticsearch.k8s.elastic.co/cluster-name
        operator: "In"
        values: ["cluster-name"]
    -  key: elasticsearch.k8s.elastic.co/statefulset-name
        operator: "In"
        # All of these statefulsets share data* roles, so they must be grouped together.
        values: ["cluster-name-es-masters", "cluster-name-es-data-warm", "cluster-name-es-data-cold"]

apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  labels:
    common.k8s.elastic.co/type: elasticsearch
    elasticsearch.k8s.elastic.co/cluster-name: cluster-name
  name: cluster-name-es-default-data_frozen
spec:
  maxUnavailable: 1
  selector:
    matchExpressions:
    -  key: elasticsearch.k8s.elastic.co/cluster-name
        operator: "In"
        values: ["cluster-name"]
    -  key: elasticsearch.k8s.elastic.co/statefulset-name
        operator: "In"
        values: ["cluster-name-es-data_frozen"]

apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  labels:
    common.k8s.elastic.co/type: elasticsearch
    elasticsearch.k8s.elastic.co/cluster-name: cluster-name
  name: cluster-name-es-default-coord
spec:
  maxUnavailable: 1
  selector:
    matchExpressions:
    -  key: elasticsearch.k8s.elastic.co/cluster-name
        operator: "In"
        values: ["cluster-name"]
    -  key: elasticsearch.k8s.elastic.co/statefulset-name
        operator: "In"
        values: ["cluster-name-es-coordinating"]
```

## Implementation notes for review

I admittedly went back and forth on the implementation, specifically the method of grouping statefulSets by their associated roles. Initially when thinking through how I would group these nodeSets together I logically land on a recursive-style algorithm (traverse all sts' roles, and find all other sts' with this role, and traverse that sts' roles), which I don't want to bring into this code base, so I began investigating other potential options for this grouping. I seemed to recall something along the lines of connected points on a graph, which led me to the current implementation which is building an adjacency list, then using depth first search to build the connected "groups" of statefulsets. I keep feeling that there's a simpler method of doing this grouping, so suggestions are *strongly* welcome in this area.

### TODO

- [ ] full testing in a cluster
- [x] unit tests
- [ ] Documentation (do we want docs in this PR, or in an additional?)

prodsecmachine commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/elastic-cloud-on-k8s/pr-checks/88ed15a1-fbd2-451e-b6dd-fcde827fa73d?refs=6d70499bcdf120c21ad3a3403c6044e9534a101f&source=prChecksComment)

naemono commented on pkg/controller/elasticsearch/pdb/dfs.go:
```diff
@@ -0,0 +1,153 @@
+	// Populate the adjacency list with each StatefulSet index, and the slice of StatefulSet
+	// indices which share roles.
+	for _, indices := range roleToIndices {
+		for i := 1; i < len(indices); i++ {
+			adjList[indices[0]] = append(adjList[indices[0]], indices[i])
+			adjList[indices[i]] = append(adjList[indices[i]], indices[0])
+			for j := 1; j < len(indices); j++ {
+				if indices[i] != indices[j] && !slices.Contains(adjList[indices[i]], indices[j]) {
+					adjList[indices[i]] = append(adjList[indices[i]], indices[j])
+					adjList[indices[j]] = append(adjList[indices[j]], indices[i])
+				}
+			}
+		}
```
I'll investigate this. It looks like the same nested loop to me initially.

naemono commented:
buildkite test this

pebrc commented on pkg/controller/elasticsearch/pdb/dfs.go:
```diff
@@ -0,0 +1,153 @@
+// is normalized to the same data role.
+func normalizeRole(role string) string {
+	if slices.Contains(dataRoles, role) {
+		return string(esv1.DataRole)
```
I think we should only do that if the user has specified the generic data role. If they use the tiered data roles we should keep them and created dedicated PDBs. The rationale being that with tiered data roles indices are very likely to be fully within one of the tiers. For example for the the hot tier both primary and replica shards are in the hot tier. So the tier should be considered as an logical availability domain and be protected via a dedicated PDB. If a a node in the hot tier is disrupted and we are in yellow health then further disruptions in let's say the cold data tier won't make the situation worse.

pebrc commented on pkg/controller/elasticsearch/pdb/roles.go:
```diff
@@ -0,0 +1,395 @@
+
+		// Create a PDB for this group
+		//
+		// TODO: Remove before merge: It feels like there's a possibility of overlapping pdb names here.
```
If we build the PDBs based on the predefined list of priorities then PDB names are more or less fixed and this problem goes away.

naemono commented on pkg/controller/elasticsearch/pdb/reconcile.go:
```diff
@@ -170,7 +208,7 @@ func buildPDBSpec(es esv1.Elasticsearch, statefulSets sset.StatefulSetList) poli
 	nodeCount := statefulSets.ExpectedNodeCount()
 	// maybe allow some Pods to be disrupted
-	minAvailable := nodeCount - allowedDisruptions(es, statefulSets)
+	minAvailable := nodeCount - allowedDisruptionsForRole(es, esv1.DataRole, statefulSets)
```
I was trying to keep things DRY from a logic standpoint across both implementations, but since your question [here](https://github.com/elastic/cloud-on-k8s/pull/8780#discussion_r2249330421), I'm fully on-board with not mixing this logic. Thanks for catching that!

naemono commented on pkg/controller/elasticsearch/label/label.go:
```diff
@@ -85,7 +85,7 @@ func IsMasterNodeSet(statefulSet appsv1.StatefulSet) bool {
 // IsDataNodeSet returns true if the given StatefulSet specifies data nodes.
 func IsDataNodeSet(statefulSet appsv1.StatefulSet) bool {
-	return NodeTypesDataLabelName.HasValue(true, statefulSet.Spec.Template.Labels)
+	return NodeTypesDataLabelName.HasValue(true, statefulSet.Spec.Template.Labels) || NodeTypesDataHotLabelName.HasValue(true, statefulSet.Spec.Template.Labels) || NodeTypesDataColdLabelName.HasValue(true, statefulSet.Spec.Template.Labels) || NodeTypesDataContentLabelName.HasValue(true, statefulSet.Spec.Template.Labels) || NodeTypesDataWarmLabelName.HasValue(true, statefulSet.Spec.Template.Labels)
```
This was required as these other data* node role types should be considered in this function, right?

barkbay commented on pkg/controller/elasticsearch/pdb/roles.go:
```diff
@@ -0,0 +1,518 @@
+}
+
+// reconcileRoleSpecificPDBs creates and reconciles PodDisruptionBudgets per nodeSet roles for enterprise-licensed clusters.
+func reconcileRoleSpecificPDBs(
```
nit, we are reconciling PDBs here while this file is named `roles.go`, and we have another file named `reconcile.go`.

I wonder if this function should go in `reconcile.go`? Or maybe we should have `reconcile_default.go` and `reconcile_roles_specific.go` (or something along those lines)?

naemono replied:
I went with `reconcile_default.go` and `reconcile_with_roles.go`.. Lmk if this works for you.

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile.go:
```diff
@@ -19,17 +20,44 @@ import (
+//
 // If the spec has disabled the default PDB, it will ensure none exist.
 func Reconcile(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch, statefulSets sset.StatefulSetList, meta metadata.Metadata) error {
+	licenseChecker := lic.NewLicenseChecker(k8sClient, es.Namespace)
```
The second parameter must be the operator namespace (not the Elasticsearch one).

barkbay replied:
I think we are still using the Elasticsearch namespace here instead of the controller namespace: https://github.com/elastic/cloud-on-k8s/pull/8780/files#diff-7166470f6fb1603354bea209027f74680de6e68f66ea9979952b6e8ebed8fdccR41

barkbay replied:
I think this is not fixed (or I'm missing something)

barkbay reviewed (commented):
I'm still going through the code.

barkbay commented on pkg/controller/elasticsearch/pdb/roles.go:
```diff
@@ -0,0 +1,518 @@
+		return deleteAllRoleSpecificPDBs(ctx, k8sClient, es)
+	}
+
+	// Always ensure any existing default PDB is removed
+	if err := deleteDefaultPDB(ctx, k8sClient, es); err != nil {
+		return fmt.Errorf("while deleting the default PDB: %w", err)
+	}
```
Should we not do that only once the others PDBs have been successfully reconciled?

barkbay reviewed (commented):
Not sure why we are not already watching PDBs 🤔 

```diff
--- a/pkg/controller/elasticsearch/elasticsearch_controller.go
+++ b/pkg/controller/elasticsearch/elasticsearch_controller.go
@@ -13,6 +13,7 @@ import (
        "go.elastic.co/apm/v2"
        appsv1 "k8s.io/api/apps/v1"
        corev1 "k8s.io/api/core/v1"
+       policyv1 "k8s.io/api/policy/v1"
        apierrors "k8s.io/apimachinery/pkg/api/errors"
        "k8s.io/apimachinery/pkg/types"
        "k8s.io/client-go/tools/record"
@@ -92,6 +93,12 @@ func addWatches(mgr manager.Manager, c controller.Controller, r *ReconcileElasti
                return err
        }
 
+       // Watch PodDisruption Budgets
+       if err := c.Watch(
+               source.Kind(mgr.GetCache(), &policyv1.PodDisruptionBudget{}, handler.TypedEnqueueRequestForOwner[*policyv1.PodDisruptionBudget](mgr.GetScheme(), mgr.GetRESTMapper(), &esv1.Elasticsearch{}, handler.OnlyControllerOwner()))); err != nil {
+               return err
+       }
+
        // Watch pods belonging to ES clusters
        if err := watches.WatchPods(mgr, c, label.ClusterNameLabelName); err != nil {
                return err
```

barkbay commented on pkg/controller/elasticsearch/pdb/roles.go:
```diff
@@ -0,0 +1,518 @@
+var (
+	// group the statefulsets by the priority of their roles.
+	// master, data_*, ingest, ml, transform, coordinating, and we ignore remote_cluster_client as it has no impact on availability
+	priority = []string{"master", "data", "data_frozen", "ingest", "ml", "transform", "coordinating"}
```
Should we try to reuse the types and logic from [https://github.com/elastic/cloud-on-k8s/blob/main/pkg/apis/elasticsearch/v1/elasticsearch_config.go](pkg/apis/elasticsearch/v1/elasticsearch_config.go)? It feels like there is a bit a of duplicated logic in some places, for example in `normalizeRole`  which is almost already implemented in [CanContainData()](https://github.com/elastic/cloud-on-k8s/blob/1832f4bbc4e280ffeff5df96fd4ce3bf503899b4/pkg/apis/elasticsearch/v1/elasticsearch_config.go#L62)

```suggestion
	priority = []esv1.NodeRole{esv1.MasterRole, esv1.DataRole, esv1.DataFrozenRole, esv1.IngestRole, esv1.MLRole, esv1.TransformRole, esv1.CoordinatingRole}
```

barkbay commented on pkg/controller/elasticsearch/pdb/roles.go:
```diff
@@ -0,0 +1,518 @@
+	return roleSpecificPDBs, nil
+}
+
+// deleteAllRoleSpecificPDBs deletes all existing role-specific PDBs for the cluster by retrieving
+// all PDBs in the namespace with the cluster label and verifying the owner reference.
+func deleteAllRoleSpecificPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch) error {
```
```suggestion
// deleteAllPDBs deletes all existing PDBs for the cluster by retrieving
// all PDBs in the namespace with the cluster label and verifying the owner reference.
func deleteAllPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch) error {
```

Or maybe I'm missing how we are targeting "role specific" PDBs.

naemono replied:
Ok, I'm handling both v1 and v1alpha1 logic now, but it has made the code a bit ugly. Lmk if you see optimizations.

barkbay commented on pkg/controller/elasticsearch/pdb/roles.go:
```diff
@@ -0,0 +1,518 @@
+	// Delete only PDBs that are owned by this Elasticsearch controller
+	for _, pdb := range pdbList.Items {
+		// Check if this PDB is owned by the Elasticsearch resource
+		if isOwnedByElasticsearch(pdb, es) {
+			if err := k8sClient.Delete(ctx, &pdb); err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// isOwnedByElasticsearch checks if a PDB is owned by the given Elasticsearch resource.
+func isOwnedByElasticsearch(pdb policyv1.PodDisruptionBudget, es esv1.Elasticsearch) bool {
+	for _, ownerRef := range pdb.OwnerReferences {
+		if ownerRef.Controller != nil && *ownerRef.Controller &&
+			ownerRef.APIVersion == esv1.GroupVersion.String() &&
+			ownerRef.Kind == esv1.Kind &&
+			ownerRef.Name == es.Name {
+			return true
+		}
+	}
+	return false
+}
```
```suggestion
		if k8s.HasOwner(&pdb, &es) {
			if err := k8sClient.Delete(ctx, &pdb); err != nil && !apierrors.IsNotFound(err) {
				return err
			}
		}
	}
	return nil
}
```

barkbay commented on pkg/controller/elasticsearch/pdb/roles.go:
```diff
@@ -0,0 +1,518 @@
+	return false
+}
+
+// podDisruptionBudgetName returns the name of the PDB.
+func podDisruptionBudgetName(esName string, role string) string {
```
Let's maybe move this somewhere around https://github.com/elastic/cloud-on-k8s/blob/1832f4bbc4e280ffeff5df96fd4ce3bf503899b4/pkg/apis/elasticsearch/v1/name.go#L177 ?

naemono commented on pkg/controller/elasticsearch/pdb/reconcile.go:
```diff
@@ -19,17 +20,44 @@ import (
+//
+//	by grouping statefulSets by associated Elasticsearch node roles into the same PDB, and then dynamically maxUnavailable
+//	according to whatever cluster health is optimal for the set of roles.
+//
```
My formatter was forcing this. I've restructured it.

pebrc commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,560 @@
+	if err != nil {
+		return nil, err
+	}
+	nodeRoles := make([]esv1.NodeRole, len(cfg.Node.Roles))
+	for i, role := range cfg.Node.Roles {
+		nodeRoles[i] = esv1.NodeRole(role)
+	}
+	return nodeRoles, nil
```
I don't think this is correct. By default not specifying any roles means the node has all roles. We have logic implemented already that takes that into account when calculating the labels for the statefulset. 
With this logic here we would treat a nodeSet without explicit roles incorrectly as coordinating. However the only way to create a coordinating only nodeSet is by setting `node.roles: []`

naemono replied:
Good catch here. This is a wrinkle I wasn't aware of, tbh. I've fixed the issue, and added additional unit tests to ensure this is handled properly.

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,560 @@
+// deleteAllRoleSpecificPDBs deletes all existing role-specific PDBs for the cluster by retrieving
+// all PDBs in the namespace with the cluster label and verifying the owner reference.
+func deleteAllRoleSpecificPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch) error {
+	v1Available, err := isPDBV1Available(k8sClient)
```
I think this comes from my comment here: https://github.com/elastic/cloud-on-k8s/pull/8780#discussion_r2256543482 It feels a bit odd to have different ways to handle the pdb API. `policy/v1` is available since `1.21` so I think we can just get rid of that logic?

naemono replied:
```
Supported versions:

*  Kubernetes 1.29-1.33
```

We should absolutely remove this logic. Great point. This will clean up the old and new code nicely.

naemono commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,560 @@
+		return 1
+	}
+	// There's a risk the single master of the cluster gets removed, don't allow it.
+	if role == esv1.MasterRole && statefulSets.ExpectedMasterNodesCount() == 1 {
```
This took some additional functional changes to resolve, but I believe I resolved this for both master and ingest decision branches, and added additional unit tests to verify.

naemono commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,560 @@
+// allowedDisruptionsForRole returns the maximum number of pods that can be disrupted for a given role.
+func allowedDisruptionsForRole(
+	es esv1.Elasticsearch,
+	role esv1.NodeRole,
```
Yes, I think this is the case *in certain scenarios*, which I recently made changes to cover. I'd like your input on whether I've covered the needed cases though @pebrc , thanks.

pebrc commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles_test.go:
```diff
@@ -977,6 +860,24 @@ func Test_allowedDisruptionsForRole(t *testing.T) {
+			args: args{
+				es:              esv1.Elasticsearch{Status: esv1.ElasticsearchStatus{Health: esv1.ElasticsearchGreenHealth}},
+				role:            []esv1.NodeRole{esv1.DataRole},
+				allStatefulSets: sset.StatefulSetList{ssetfixtures.TestSset{Replicas: 3}.Build()},
```
For data nodes 2 can already be considered HA: one for the primary one for the replica

pebrc commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles_test.go:
```diff
@@ -1014,38 +915,43 @@ func Test_allowedDisruptionsForRole(t *testing.T) {
+			want: 1,
+		},
+		{
+			name: "green health but only 1 master: 1 disruption allowed as the cluster is non-ha",
```
I think test description does not match the test here:
* test is for data roles
* data tier is actually HA
* cluster is green anyway?

pebrc commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles_test.go:
```diff
@@ -0,0 +1,1340 @@
+			want: 1,
+		},
+		{
+			name: "yellow health and only 1 data node: 0 disruption allowed for data role",
```
Based on our discussions I would expect one disruption to be allowed as with a single node data tier there can be no expectation of HA?

pebrc reviewed (dismissed):
I think we are close but I found a few more things that need to be fixed especially around BWC.

barkbay commented:
> Dismissing my review as not to block progress on this PR while I am unvailable. @barkbay will continue helping with the review.

👍 I'll catch up with this PR once the last comments/suggestions have been addressed (just to avoid duplicate findings).

naemono commented:
> > Dismissing my review as not to block progress on this PR while I am unvailable. @barkbay will continue helping with the review.
> 
> 👍 I'll catch up with this PR once the last comments/suggestions have been addressed (just to avoid duplicate findings).

@barkbay This has been updated per Peter's review. Feel free to proceed. There are open questions [here](https://github.com/elastic/cloud-on-k8s/pull/8780#discussion_r2296184877)

barkbay commented on pkg/apis/elasticsearch/v1/elasticsearch_types.go:
```diff
@@ -103,9 +103,17 @@ type ElasticsearchSpec struct {
+	// With a Basic license:
+	//   The default budget doesn't allow any Pod to be removed in case the cluster is not green or if there is only one node of type `data` or `master`.
+	//   In all other cases the default PodDisruptionBudget sets `minUnavailable` equal to the total number of nodes minus 1.
+	// With an Enterprise license:
```
```suggestion
	// With an Enterprise license and if podDisruptionBudget.spec is empty:
```

barkbay commented on pkg/apis/elasticsearch/v1/name.go:
```diff
@@ -200,3 +200,13 @@ func StackConfigAdditionalSecretName(esName string, secretName string) string {
+
+// PodDisruptionBudgetNameForRole returns the name of the PodDisruptionBudget for a given Elasticsearch cluster name and role.
+func PodDisruptionBudgetNameForRole(esName string, role string) string {
+	name := DefaultPodDisruptionBudget(esName) + "-" + role
+	// For coordinating nodes (no roles), append "coordinating" to the name
+	if role == "" {
+		name += "coordinating"
+	}
+	return name
```
I think it is the purpose of the namer to add the `-`:
```suggestion
	if role == "" {
		role = "coordinating"
	}
	return ESNamer.Suffix(esName, defaultPodDisruptionBudget, role)
```

barkbay commented on docs/reference/api-docs.md:
```diff
@@ -10,4 +10,4 @@ applies_to:
 
-This page has moved [here](./api-reference/index.md).
\ No newline at end of file
+This page has moved [here](./api-reference/index.md).
```
I don't understand what is changed here?

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles_test.go:
```diff
@@ -0,0 +1,1349 @@
+			resourcesList, err := tt.args.builder.BuildResourcesList()
+			require.NoError(t, err)
+
+			statefulSets := tt.args.builder.GetStatefulSets()
+			// get the specified statefulSet from the list to pass as argument to getRolesForStatefulSet
+			var statefulSet appsv1.StatefulSet
+			found := false
+			for _, sset := range statefulSets {
+				if sset.Name == tt.args.statefulSetName {
+					statefulSet = sset
+					found = true
+					break
+				}
+			}
```
```suggestion
			// get the specified statefulSet from the list to pass as argument to getRolesForStatefulSet
			statefulSet, found := tt.args.builder.GetStatefulSets().GetByName(tt.args.statefulSetName)
```

with

```golang

import (
	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/elasticsearch/sset"
)

func (b Builder) GetStatefulSets() sset.StatefulSetList {
	return b.StatefulSets
}
```

naemono commented:
> I think I'm almost done with the review, beside a few nits:
> 
> * I don't understand why are we not allowing a disruption if there is only 1 expected data node  : [#8780 (comment)](https://github.com/elastic/cloud-on-k8s/pull/8780#discussion_r2303751241) But I definitely feel like it's me missing something obvious 🙃

We discussed this offline, and I've added some comments to the disruptions func.

> * Maybe we need to clarify how we merge the provided PDB spec: [#8780 (comment)](https://github.com/elastic/cloud-on-k8s/pull/8780#discussion_r2303384256)

This is complete.

> * License checker seems to still be broken: [#8780 (comment)](https://github.com/elastic/cloud-on-k8s/pull/8780#discussion_r2303145422)

I had somehow missed this. It's complete as well.

> * I still have to run a test outside of my local env.
> 
> Unit test coverage seems really great! 👍 But I'm also wondering if this would deserve at least one e2e test, where a license is added and removed for example (not a strong feeling though...).

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,449 @@
+	expectedResources nodespec.ResourcesList,
+	v version.Version,
+) ([]esv1.NodeRole, error) {
+	forStatefulSet, err := expectedResources.ForStatefulSet(statefulSet.Name)
```
I think we have a bug here: if a tier is removed from the Elasticsearch spec then there is a point where no StatefulSet set is going to be expected. This is going to lead to the following error which, I believe, is going to prevent any further reconciliation:

```json
{
    "log.level": "error",
    "@timestamp": "2025-09-02T06:58:28.715Z",
    "log.logger": "manager.eck-operator",
    "message": "Reconciler error",
    "service.version": "3.2.0-SNAPSHOT+586f4ed2",
    "service.type": "eck",
    "ecs.version": "1.4.0",
    "controller": "elasticsearch-controller",
    "object": {
        "name": "elasticsearch-sample",
        "namespace": "default"
    },
    "namespace": "default",
    "name": "elasticsearch-sample",
    "reconcileID": "914b1631-3324-449b-a9e9-bd0cdea3180a",
    "error": "while retrieving expected role-specific PDBs: while grouping StatefulSets by roles: no expected resources for StatefulSet elasticsearch-sample-es-frozen",
    "errorCauses": [
        {
            "error": "while retrieving expected role-specific PDBs: while grouping StatefulSets by roles: no expected resources for StatefulSet elasticsearch-sample-es-frozen"
        }
    ],
    "error.stack_trace": "sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\t/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.21.0/pkg/internal/controller/controller.go:353\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\t/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.21.0/pkg/internal/controller/controller.go:300\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.1\n\t/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.21.0/pkg/internal/controller/controller.go:202"
}
```

In the example above I removed a frozen tier from the spec.

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,449 @@
+//     to a cluster which is already unhealthy for an unknown reason.
+//   - In a single node cluster (not highly-available) always allow 1 disruption
+//     to ensure K8s nodes operations can be performed when the cluster health is not unknown or red.
+//   - For data roles, allow 1 disruption if health is green.
```
One thing which is not obvious when reading this code is "when" frozen is handled differently. Let's maybe try to explain where the magic happens in the godoc?
```suggestion
//   - For data roles, excluding frozen which should have been filtered in groupBySharedRoles, allow 1 disruption if health is green.
```

barkbay commented on pkg/controller/elasticsearch/pdb/roles.go:
```diff
@@ -0,0 +1,518 @@
+	}
+
+	return roles
+}
```
My suggestions was definitely wrong because some `StatefulSets` may not be expected if their related node sets have been removed from the ES spec.

What I initially didn't like with the labels is that it feels we are "reverse engineering" the roles from the labels, while labels should be used to convey identity information, not to infer configuration details. Unfortunately I'm not sure there is a way around that (maybe properly serialize the roles in an annotation on each Pod?).

Something which is not clear to me though is why we decided to iterate through the existing StatefulSets? Why not iterating through the running Pods? The reason I'm asking is that when there is a config change (for example a role is added or removed) this change is not immediately reflected in the running Pods. StatefulSets are immediately updated, but Pods may still run with the previous configuration for a while, until they are deleted. Therefore inferring the roles actually deployed on a cluster from the StatefulSet feels brittle? (or I'm overthinking 😄 )

naemono replied:
I've updated this logic to take into account covering deleted/renamed STS' as well. I don't love it because now we have 2 ways to retrieve roles from the statefulsets. 

I see your note here about deducing this information from the running pods, and I'll think about this.  My initial thought is that we're (the eck operator) providing the labels to the STS' so the only way it would be "brittle" was if there were issues with the logic where we assign these labels.

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,456 @@
+		if !slices.ContainsFunc([]appsv1.StatefulSet(resources.StatefulSets()), func(s appsv1.StatefulSet) bool {
+			return s.Name == sset.Name
+		}) {
+			continue
```
If a user removes or renames a node set then the underlying StatefulSet is going to be preserved by ECK as long the Pods cannot be safely removed. For example because their data have not been migrated to other ES nodes, or because in the case of master nodes we remove them one by one. Therefore, even if they are not expected these StatefulSet should still be considered as part of the cluster until they are _eventually_ drained.

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,443 @@
+	// so we can initially assume it is, and then check for any labels that are set to true
+	// to invalidate the assumption.
+	isCoordinating := true
+	// Define label-role mappings
+	labelRoleMappings := []struct {
+		labelName string
+		role      esv1.NodeRole
+	}{
+		{string(label.NodeTypesMasterLabelName), esv1.MasterRole},
+		{string(label.NodeTypesDataLabelName), esv1.DataRole},
+		{string(label.NodeTypesIngestLabelName), esv1.IngestRole},
+		{string(label.NodeTypesMLLabelName), esv1.MLRole},
+		{string(label.NodeTypesTransformLabelName), esv1.TransformRole},
+		{string(label.NodeTypesRemoteClusterClientLabelName), esv1.RemoteClusterClientRole},
+		{string(label.NodeTypesDataHotLabelName), esv1.DataHotRole},
+		{string(label.NodeTypesDataWarmLabelName), esv1.DataWarmRole},
+		{string(label.NodeTypesDataColdLabelName), esv1.DataColdRole},
+		{string(label.NodeTypesDataContentLabelName), esv1.DataContentRole},
+		{string(label.NodeTypesDataFrozenLabelName), esv1.DataFrozenRole},
+	}
+	// Check each label-role mapping
+	for _, mapping := range labelRoleMappings {
```
Let's maybe move this where the labels are defined?

```diff
diff --git a/pkg/controller/elasticsearch/label/label.go b/pkg/controller/elasticsearch/label/label.go
index 5d4911886..62c49ad2d 100644
--- a/pkg/controller/elasticsearch/label/label.go
+++ b/pkg/controller/elasticsearch/label/label.go
@@ -59,6 +59,25 @@ const (
        Type = "elasticsearch"
 )
 
+type RoleMapping struct {
+       LabelName string
+       Role      esv1.NodeRole
+}
+
+var RoleMappings = []RoleMapping{
+       {string(NodeTypesMasterLabelName), esv1.MasterRole},
+       {string(NodeTypesDataLabelName), esv1.DataRole},
+       {string(NodeTypesIngestLabelName), esv1.IngestRole},
+       {string(NodeTypesMLLabelName), esv1.MLRole},
+       {string(NodeTypesTransformLabelName), esv1.TransformRole},
+       {string(NodeTypesRemoteClusterClientLabelName), esv1.RemoteClusterClientRole},
+       {string(NodeTypesDataHotLabelName), esv1.DataHotRole},
+       {string(NodeTypesDataWarmLabelName), esv1.DataWarmRole},
+       {string(NodeTypesDataColdLabelName), esv1.DataColdRole},
+       {string(NodeTypesDataContentLabelName), esv1.DataContentRole},
+       {string(NodeTypesDataFrozenLabelName), esv1.DataFrozenRole},
+}
+
```

```suggestion
	// Check each label-role mapping
	for _, mapping := range label.RoleMappings {
```

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,443 @@
+		// labels in the statefulSet, not the expected configuration.
+		roles := getRolesFromStatefulSet(sset)
+
+		if len(roles) == 0 {
```
IIUC `getRolesFromStatefulSet` returns an empty results only if we are in the (unexpected) situation where there is no labels on the `Pod`. In the case of coordinating node I believe we have at least `esv1.CoordinatingRole`.

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,443 @@
+	roles := []esv1.NodeRole{}
+	labels := statefulSet.Spec.Template.Labels
+	if labels == nil {
+		return roles
```
Should we return an error?

barkbay commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,443 @@
+	if es.Status.Health == esv1.ElasticsearchUnknownHealth ||
+		es.Status.Health == esv1.ElasticsearchHealth("") ||
+		es.Status.Health == esv1.ElasticsearchRedHealth {
+		return 0
```
Should we log something to let the user know why no disruption is allowed?

barkbay reviewed (approved):
LGTM 🚢 

I left a couple of suggestions and the following questions:
- [ ] Should we propagate the PDB's metadata set in the ES resource.
- [x] Check if it is interesting to add some logs when we don't allow any disruption.

naemono commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,440 @@
+		Spec: buildRoleSpecificPDBSpec(es, role, statefulSets, allStatefulSets),
+	}
+
+	mergedMeta := meta.Merge(metadata.Metadata{
+		Labels:      pdb.Labels,
+		Annotations: pdb.Annotations,
+	})
```
I totally missed this discussion 2 weeks ago buried in this PR. I'm looking now.

pebrc commented on pkg/controller/elasticsearch/pdb/reconcile_with_roles.go:
```diff
@@ -0,0 +1,440 @@
+		Spec: buildRoleSpecificPDBSpec(ctx, es, role, statefulSets, allStatefulSets),
+	}
+
+	if es.Spec.PodDisruptionBudget != nil {
+		pdb.Annotations = es.Spec.PodDisruptionBudget.ObjectMeta.Annotations
+		pdb.Labels = es.Spec.PodDisruptionBudget.ObjectMeta.Labels
+	}
```
Missing unit test for this? Also the only reason that this does not modify the original maps IIUC is that the following call `meta.Merge` does clone the maps. Maybe at least at a comment?

naemono replied:
I spoke to you about this, and have a comment that lines up with your thoughts.
